<a href="https://colab.research.google.com/github/Andrii-Oliinyk-2726/Data_Science_HW11/blob/main/HW11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import time
import numpy as np

import matplotlib.pyplot as plt
from keras.datasets import imdb
from keras.utils import pad_sequences
from keras.models import Sequential
from keras.layers import Embedding, SimpleRNN, LSTM, Bidirectional, Dense


max_features = 10000
maxlen = 500
batch_size = 32

(input_train, y_train), (input_test, y_test) = imdb.load_data(num_words=max_features)

input_train = pad_sequences(input_train, maxlen=maxlen)
input_test = pad_sequences(input_test, maxlen=maxlen)

17464789/17464789 [==============================] - 0s 0us/step


**Рекурентна нейронна мережа (RNN)**

In [2]:
model = Sequential()
model.add(Embedding(max_features, 32))
model.add(SimpleRNN(32))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, None, 32)          320000    
                                                                 
 simple_rnn (SimpleRNN)      (None, 32)                2080      
                                                                 
 dense (Dense)               (None, 1)                 33        
                                                                 
Total params: 322113 (1.23 MB)
Trainable params: 322113 (1.23 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [3]:
start_time = time.time()
history = model.fit(input_train, y_train, epochs=5, batch_size=128, validation_split=0.3)
processing_time = time.time() - start_time
print(f"\nTotal training time: {processing_time:.2f} seconds")

Epoch 1/5
137/137 [==============================] - 24s 153ms/step - loss: 0.5992 - acc: 0.6680 - val_loss: 0.4253 - val_acc: 0.8159
Epoch 2/5
137/137 [==============================] - 22s 158ms/step - loss: 0.3889 - acc: 0.8357 - val_loss: 0.4517 - val_acc: 0.8164
Epoch 3/5
137/137 [==============================] - 20s 148ms/step - loss: 0.3070 - acc: 0.8752 - val_loss: 0.3706 - val_acc: 0.8423
Epoch 4/5
137/137 [==============================] - 21s 150ms/step - loss: 0.2523 - acc: 0.9003 - val_loss: 0.4282 - val_acc: 0.8068
Epoch 5/5
137/137 [==============================] - 22s 163ms/step - loss: 0.1968 - acc: 0.9262 - val_loss: 0.4379 - val_acc: 0.8069

Total training time: 145.26 seconds


In [4]:
predictions = model.predict(input_test)
binary_predictions = [1 if pred > 0.5 else 0 for pred in predictions]
test_loss, test_acc = model.evaluate(input_test, y_test)
print('Test Loss: {}'.format(test_loss))
print('Test Accuracy: {}'.format(test_acc))

782/782 [==============================] - 18s 22ms/step - loss: 0.4454 - acc: 0.8068
Test Loss: 0.4454036355018616
Test Accuracy: 0.8068400025367737


**LSTM**

In [5]:
model = Sequential()
model.add(Embedding(max_features, 32))
model.add(LSTM(32))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_1 (Embedding)     (None, None, 32)          320000    
                                                                 
 lstm (LSTM)                 (None, 32)                8320      
                                                                 
 dense_1 (Dense)             (None, 1)                 33        
                                                                 
Total params: 328353 (1.25 MB)
Trainable params: 328353 (1.25 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [6]:
start_time = time.time()
history = model.fit(input_train, y_train, epochs=5, batch_size=128, validation_split=0.3)
processing_time = time.time() - start_time
print(f"\nTotal training time: {processing_time:.2f} seconds")


Epoch 1/5
137/137 [==============================] - 47s 331ms/step - loss: 0.6140 - acc: 0.6609 - val_loss: 0.4361 - val_acc: 0.8200
Epoch 2/5
137/137 [==============================] - 44s 324ms/step - loss: 0.3734 - acc: 0.8474 - val_loss: 0.3393 - val_acc: 0.8600
Epoch 3/5
137/137 [==============================] - 46s 340ms/step - loss: 0.2816 - acc: 0.8911 - val_loss: 0.3516 - val_acc: 0.8519
Epoch 4/5
137/137 [==============================] - 50s 366ms/step - loss: 0.2354 - acc: 0.9098 - val_loss: 0.2885 - val_acc: 0.8857
Epoch 5/5
137/137 [==============================] - 50s 365ms/step - loss: 0.2049 - acc: 0.9275 - val_loss: 0.3041 - val_acc: 0.8843

Total training time: 263.84 seconds


In [7]:
predictions = model.predict(input_test)
test_loss, test_acc = model.evaluate(input_test, y_test)
print('Test Loss: {}'.format(test_loss))
print('Test Accuracy: {}'.format(test_acc))

782/782 [==============================] - 35s 45ms/step - loss: 0.3189 - acc: 0.8775
Test Loss: 0.3189212381839752
Test Accuracy: 0.8774799704551697


**Bidirectional_LSTM**

In [8]:
model = Sequential()
model.add(Embedding(max_features, 32))
model.add(Bidirectional(LSTM(32)))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, None, 32)          320000    
                                                                 
 bidirectional (Bidirection  (None, 64)                16640     
 al)                                                             
                                                                 
 dense_2 (Dense)             (None, 1)                 65        
                                                                 
Total params: 336705 (1.28 MB)
Trainable params: 336705 (1.28 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [9]:
start_time = time.time()
history = model.fit(input_train, y_train, epochs=5, batch_size=128, validation_split=0.3)
processing_time = time.time() - start_time
print(f"\nTotal training time: {processing_time:.2f} seconds")

Epoch 1/5
137/137 [==============================] - 96s 668ms/step - loss: 0.6502 - acc: 0.6066 - val_loss: 0.5036 - val_acc: 0.8023
Epoch 2/5
137/137 [==============================] - 82s 602ms/step - loss: 0.4551 - acc: 0.8037 - val_loss: 0.3602 - val_acc: 0.8569
Epoch 3/5
137/137 [==============================] - 83s 607ms/step - loss: 0.3171 - acc: 0.8744 - val_loss: 0.3527 - val_acc: 0.8520
Epoch 4/5
137/137 [==============================] - 82s 599ms/step - loss: 0.2579 - acc: 0.9031 - val_loss: 0.3026 - val_acc: 0.8759
Epoch 5/5
137/137 [==============================] - 84s 616ms/step - loss: 0.2296 - acc: 0.9171 - val_loss: 0.3106 - val_acc: 0.8805

Total training time: 445.92 seconds


In [10]:
predictions = model.predict(input_test)
test_loss, test_acc = model.evaluate(input_test, y_test)
print('Test Loss: {}'.format(test_loss))
print('Test Accuracy: {}'.format(test_acc))

782/782 [==============================] - 53s 67ms/step - loss: 0.3265 - acc: 0.8716
Test Loss: 0.32651400566101074
Test Accuracy: 0.8715599775314331


**Deep_LSTM**

In [11]:
model = Sequential()
model.add(Embedding(max_features, 32))
model.add(Bidirectional(LSTM(32, return_sequences=True)))
model.add(Bidirectional(LSTM(32)))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_3 (Embedding)     (None, None, 32)          320000    
                                                                 
 bidirectional_1 (Bidirecti  (None, None, 64)          16640     
 onal)                                                           
                                                                 
 bidirectional_2 (Bidirecti  (None, 64)                24832     
 onal)                                                           
                                                                 
 dense_3 (Dense)             (None, 1)                 65        
                                                                 
Total params: 361537 (1.38 MB)
Trainable params: 361537 (1.38 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [12]:
start_time = time.time()
history = model.fit(input_train, y_train, epochs=5, batch_size=128, validation_split=0.3)
processing_time = time.time() - start_time
print(f"\nTotal training time: {processing_time:.2f} seconds")

Epoch 1/5
137/137 [==============================] - 194s 1s/step - loss: 0.6470 - acc: 0.5904 - val_loss: 0.4801 - val_acc: 0.7772
Epoch 2/5
137/137 [==============================] - 184s 1s/step - loss: 0.3924 - acc: 0.8335 - val_loss: 0.3452 - val_acc: 0.8608
Epoch 3/5
137/137 [==============================] - 182s 1s/step - loss: 0.2843 - acc: 0.8874 - val_loss: 0.3194 - val_acc: 0.8775
Epoch 4/5
137/137 [==============================] - 184s 1s/step - loss: 0.2277 - acc: 0.9132 - val_loss: 0.3180 - val_acc: 0.8684
Epoch 5/5
137/137 [==============================] - 185s 1s/step - loss: 0.1962 - acc: 0.9293 - val_loss: 0.3758 - val_acc: 0.8584

Total training time: 988.55 seconds


In [13]:
predictions = model.predict(input_test)
test_loss, test_acc = model.evaluate(input_test, y_test)
print('Test Loss: {}'.format(test_loss))
print('Test Accuracy: {}'.format(test_acc))

782/782 [==============================] - 111s 142ms/step - loss: 0.3832 - acc: 0.8537
Test Loss: 0.3832310438156128
Test Accuracy: 0.8536800146102905


Проведення кількох запусків моделей з приблизно однаковими параметрами показали що RNN на навчання витратила 145 секунд (Test Accuracy: 0.8068),  LSTM - 263 секунди (Test Accuracy: 0.8774), Bidirectional_LSTM - 445 секунд (Test Accuracy: 0.8715) та Deep_LSTM - 988 секунд (Test Accuracy: 0.8536).

Вибір конкретної архітектури має залежати від конкретної задачі а також від наявності ресурсів для побудови моделі.